iothub/device/src/InternalClient.cs

```diff
@@ -1273,6 +1290,130 @@ public Task<Message> ReceiveAsync(CancellationToken cancellationToken)
             }
         }
 
+        /// <summary>
+        /// Registers a new delegate for receiving a message from the device queue using the default timeout.
+        /// If a delegate is already registered it will be replaced with the new delegate.
+        /// If a null delegate is passed, it will disable triggering any callback on receiving messages from the service.
+        /// <param name="messageHandler">The delegate to be used when a could to device message is received by the client.</param>
+        /// <param name="userContext">Generic parameter to be interpreted by the client code.</param>
+        /// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+        /// </summary>
+        /// <returns>The task containing the event</returns>
+        public async Task SetReceiveMessageHandlerAsync(ReceiveMessageCallback messageHandler, object userContext, CancellationToken cancellationToken)
+        {
+            try
+            {
+                if (Logging.IsEnabled)
+                {
+                    Logging.Enter(this, messageHandler, userContext, nameof(SetReceiveMessageHandlerAsync));
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
```

Comment:
Well, kind-of yes and no;.
Yes, because yeah, we should not perform big chunks of operations and then only check for cancellations at the beginning.
No, because then we'd be calling it after every other line. 

The way I see it, the Tasks should be broken into logical sections that can be executed as a single block, and then when you enter a method, you check if the cancellation token has been cancelled, if not, then perform the next step of operations.

Reply:
For this scenario, we check if cancellation has been requested, if not, then call enable/disable the c2d subscription, and in there, delegate the cancellation token check to the pipeline/ transport handler.